docs/rely/mock-functions.md

```diff
@@ -0,0 +1,209 @@
+
+## Using Mock Functions
+
+### Creating Mocks
+
+There are constructors for creating mocks that take up to 7 arguments. A default implementation must be specified.
+
+```reason
+open TestFramework;
+
+let calculateInterest = Mock.mock3((p, r, t) => p *. exp(r *. t));
+```
+
+```reason
+module Mock: {
+  type t('fn, 'returnType, 'tupledArgs);
+
+  let mock1: ('args => 'ret) => t('args => 'ret, 'ret, 'args);
+
+  let mock2:
+    (('arg1, 'arg2) => 'ret) =>
+    t(('arg1, 'arg2) => 'ret, 'ret, ('arg1, 'arg2));
+
+  let mock3:
+    (('arg1, 'arg2, 'arg3) => 'ret) =>
+    t(('arg1, 'arg2, 'arg3) => 'ret, 'ret, ('arg1, 'arg2, 'arg3));
+
+  let mock4:
+    (('arg1, 'arg2, 'arg3, 'arg4) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4),
+    );
+
+  let mock5:
+    (('arg1, 'arg2, 'arg3, 'arg4, 'arg5) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5),
+    );
+
+  let mock6:
+    (('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6),
+    );
+
+  let mock7:
+    (('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6, 'arg7) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6, 'arg7) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6, 'arg7),
+    );
+};
+```
+
+### Calling Mock Functions
+
+The `Mock.mockN` constructor returns a mutable opaque type for tracking call information. To access the underlying function, use `Mock.fn`.
+
+```reason
+open TestFramework;
+
+describe("Mock documentation", ({test}) => {
+  test("calling a mock function", ({expect}) => {
+    let mockDouble = Mock.mock1(x => x * 2);
+    let mockFunction = Mock.fn(mockDouble);
+
+    expect.int(mockFunction(2)).toBe(4);
+  })
+});
+```
+
+### Injecting via Functor
+
+When testing code with side effects such as talking to a database or making network requests, it is often desirable to be able to mock out these side effects. A straightforward way of achieving this is by passing the side effect performing functions as arguments, however this is not always desirable due to the problem of threading such arguments through deeply nested code. The use of module functions (functors) can accomplish a similar thing and somewhat alleviate this issue. 
+
+For example suppose that we have some code that uses a logger. In production we want the logger to make HTTP requests to some endpoint, however we deem this behavior undesirable for our test.
+
+Without abstracting away the dependendency on our http logger, our code might look like this.
+#### MyModule.re
+
+```reasonml
+module MyApp {
```

Comment:
Does this need to be `MyApp = {`

Reply:
Yup, also needed some semi colons and the space